### ISSUE TITLE
Strict ordering of messages leads to only some of the websocket messages getting stuck and unread

### ISSUE THREAD

**Issue by yatishb** [django/channels#393]
I'm currently running Channels@0.17.2 and faced an issue when using the strict ordering of messages through the `@enforce_ordering` decorator. I've noticed that when multiple messages were sent to my application on a single websocket, the strict ordering restriction seemed to cause some of the messages to get stuck and consequently unread.

The instruction sequence between lines L96-L117 of the `enforce_ordering` decorator (get next_order from session, test it against current message order, process message, increment next_order in session) is safe against concurrent message processing and ensures proper ordering.

Race condition issues might arise if `session.save()` is called again in the caller function (`finally` in the `channel_session` decorator) at line L66-68. It will then try to save again the same next_order in session while a subsequent message is being processed by another worker.

Improper ordering of the `session.save()` in the caller function can result in an incorrect `__channels_next_order` in the session and cause subsequent messages to get stuck.

Here's how I've been able to reproduce this issue:
* A couple of messages sent on the websocket. At this time, the message with order 0 would be picked up by the worker 1 and consumed.
* Upon the first message consumption, worker 1 would increment `__channels_next_order` to 1 and save the session within the `enforce_ordering` decorator.
* Worker 1 would continue to requeue all messages in the `__wait__` channel and return to the caller function (`@channel_session`) to save `channel_session` again with the next_order=1.
* At this time, it so happens that worker 1 gets stuck attempting to save the session.
* Another worker 2, now picks up the second message with order 1. It reads the `next_order=1` and proceeds with consumption.
* Upon consumption of the second message, worker 2 proceeds to increment next_order to 2 and successfully saves the session.
* Sometimes it so happens that worker 1 gets unstuck at this point and re-writes the channel_session with an incorrect next_order of 1. This causes an issue when third message (with order 2) can no longer be read as all workers are expecting a message with order 1.

**Comment by andrewgodwin** [comment]
Yes, it is not a perfect solution, since anything with distributed locking is really hard to solve without a dedicated lock store.

Some changes in current Channels master help a bit - they remove the need for a `slight` mode in enforce_ordering because `websocket.connect` now blocks the socket from handshaking till it returns, and messages are now only sent after a consumer exits, so things are more atomic.

The overall problem remains, however. I'm not sure of what the right solution is given the current method of using the Django session store.

**Comment by yatishb** [comment]
Here's a solution that I came up with for this problem: https://github.com/django/channels/compare/master...yatishb:fix/enforce_ordering_handshake. 

Ideally we'd want an ability to tell whether a Django session was dirty before performing the save. I've tried to exploit the `modified` flag in the session for this purpose. What do you think about it?

**Comment by andrewgodwin** [comment]
What is the diff accomplishing - trying to prevent a second save once the method exits and gets to the outbound leg of the channel session decorator? Did you measure the difference?

**Comment by yatishb** [comment]
I'm trying to use the `modified` flag in the session store as a `dirty` indicator to know whether or not the session needs to be stored. By saving inside `enforce_ordering` and setting the flag back to `False`, as you mentioned, am preventing the channel_session outbound `finally` leg re-trying to save the already saved session and potentially over-writing a save done by any worker since.

**Comment by andrewgodwin** [comment]
Yes, it makes sense and I don't see it causing any backwards-compat issues - I'll merge it in.

**Comment by andrewgodwin** [comment]
Actually, I forgot it wasn't a PR yet - could you make a pull request for that change?

**Comment by laxmanverma** [comment]
How to do enforce_ordering in Django channels 2.2.0?
